gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -301,8 +303,10 @@ public GoogleCloudStorageImpl(
    * @param storage {@link Storage} to use for I/O.
    */
   @SuppressWarnings("CheckReturnValue")
+  @VisibleForTesting
   public GoogleCloudStorageImpl(GoogleCloudStorageOptions options, Storage storage) {
-    this(options, storage, /* downscopedAccessTokenFn= */ null);
+    this(options, storage, null, tryGetCredentialsFromStorage(storage));
+    warnIfTracingEnabled(storageOptions.isTraceLogEnabled());
```

Comment:
It seems that logic in the constructors of this class becomes more and more complicated and hard to follow. I think that we should bite a bullet and refactor it by using Builder class (or possible factory methods, but this seems less optimal) to provide input parameters for this class initialization instead of maintaining multiple contradictory constructors.